# Exercise 1 - Attendance Data

### 1) Load the attendance.csv 

In [3]:
import pandas as pd
import numpy as np

In [145]:
df_attendance = pd.read_csv('untidy_data/attendance.csv')
df_attendance = df_attendance.rename(columns={'Unnamed: 0':'name'})
df_attendance

name 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05 2018-01-06  \
0  Sally          P          T          T          H          P          A   
1   Jane          A          P          T          T          T          T   
2  Billy          A          T          A          A          H          T   
3   John          P          T          H          P          P          T   

  2018-01-07 2018-01-08  
0          T          T  
1          A          T  
2          P          T  
3          P          P

In [164]:
df_attendance.shape

(4, 9)

### 2) file and calculate an attendnace percentage for each student. 
- Note: One half day is worth 50% of a full day, and 10 tardies is equal to one absence.

In [146]:
df_attendance_melt = df_attendance.melt(id_vars='name', var_name='date', value_name='attendance')
df_attendance_melt.head()

name        date attendance
0  Sally  2018-01-01          P
1   Jane  2018-01-01          A
2  Billy  2018-01-01          A
3   John  2018-01-01          P
4  Sally  2018-01-02          T

In [147]:
df_attendance_melt.shape

(32, 3)

In [161]:
df_attendence_c = df_attendance_melt.replace({'P':1, 'A':0, 'H':.5, 'T':.9})
df_attendence_c.head()

name        date  attendance
0  Sally  2018-01-01         1.0
1   Jane  2018-01-01         0.0
2  Billy  2018-01-01         0.0
3   John  2018-01-01         1.0
4  Sally  2018-01-02         0.9

In [163]:
df_attendence_c.groupby('name').mean()

attendance
name             
Billy      0.5250
Jane       0.6875
John       0.9125
Sally      0.7625

# Exercise 2 - Coffee Levels

### 1) Read the coffee_levels.csv file.

In [168]:
df_coffee = pd.read_csv('untidy_data/coffee_levels.csv')

In [69]:
df_coffee.head()

hour coffee_carafe  coffee_amount
0     8             x       0.816164
1     9             x       0.451018
2    10             x       0.843279
3    11             x       0.335533
4    12             x       0.898291

In [70]:
df_coffee.shape

(30, 3)

### 2) Transform the data so that each carafe is in it's own column.

In [71]:
df_coffee.coffee_carafe.value_counts()

z    10
y    10
x    10
Name: coffee_carafe, dtype: int64

In [177]:
coffee_melt = df_coffee.pivot(index='hour', columns='coffee_carafe')
coffee_melt.columns = coffee_melt.columns.droplevel()
coffee_melt.columns.name = ' '
coffee_melt

x         y         z
hour                              
8     0.816164  0.189297  0.999264
9     0.451018  0.521502  0.915990
10    0.843279  0.023163  0.144928
11    0.335533  0.235529  0.311495
12    0.898291  0.017009  0.771947
13    0.310711  0.997464  0.398520
14    0.507288  0.058361  0.864464
15    0.215043  0.144644  0.436364
16    0.183891  0.544676  0.280621
17    0.391560  0.594126  0.436677

In [78]:
coffee_melt.shape

(10, 3)

### 3) Is this the best shape for the data?

#### yes, since each column is a variable, each row is an observation, and each cell is a single value

# Exercise 3 - Cake Recipes

### 1) Read the cake_recipes.csv data. 
#### Note) This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures.

In [99]:
df_cake = pd.read_csv('untidy_data/cake_recipes.csv')
df_cake

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084
5           c:top  71.306308  82.795477  92.098049  53.960273
6        d:bottom  52.799753  58.670419  51.747686  56.183110
7           d:top  96.873178  76.101363  59.571620  50.971626

In [100]:
df_cake.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8 entries, 0 to 7
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   recipe:position  8 non-null      object 
 1   225              8 non-null      float64
 2   250              8 non-null      float64
 3   275              8 non-null      float64
 4   300              8 non-null      float64
dtypes: float64(4), object(1)
memory usage: 448.0+ bytes


### 2) Tidy the data as necessary.

In [105]:
# split column 'recipe:position' to two columns for recipe and rack position
df_cake[['recipe','rack_position']] = df_cake['recipe:position'].str.split(':', expand=True)
df_cake

recipe:position        225        250        275        300 recipe  \
0        a:bottom  61.738655  53.912627  74.414730  98.786784      a   
1           a:top  51.709751  52.009735  68.576858  50.228470      a   
2        b:bottom  57.095320  61.904369  61.196980  99.248541      b   
3           b:top  82.455004  95.224151  98.594881  58.169349      b   
4        c:bottom  96.470207  52.001358  92.893227  65.473084      c   
5           c:top  71.306308  82.795477  92.098049  53.960273      c   
6        d:bottom  52.799753  58.670419  51.747686  56.183110      d   
7           d:top  96.873178  76.101363  59.571620  50.971626      d   

  rack_position  
0        bottom  
1           top  
2        bottom  
3           top  
4        bottom  
5           top  
6        bottom  
7           top

In [107]:
# drop column 'recipe:position'
df_cake = df_cake.drop(columns='recipe:position')
df_cake

225        250        275        300 recipe rack_position
0  61.738655  53.912627  74.414730  98.786784      a        bottom
1  51.709751  52.009735  68.576858  50.228470      a           top
2  57.095320  61.904369  61.196980  99.248541      b        bottom
3  82.455004  95.224151  98.594881  58.169349      b           top
4  96.470207  52.001358  92.893227  65.473084      c        bottom
5  71.306308  82.795477  92.098049  53.960273      c           top
6  52.799753  58.670419  51.747686  56.183110      d        bottom
7  96.873178  76.101363  59.571620  50.971626      d           top

In [144]:
# melt columns for different tempuratures
df_cake_melt = df_cake.melt(id_vars=['recipe','rack_position'],
                            var_name='temp',
                            value_name='taste_score')
df_cake_melt.head()

recipe rack_position temp  taste_score
0      a        bottom  225    61.738655
1      a           top  225    51.709751
2      b        bottom  225    57.095320
3      b           top  225    82.455004
4      c        bottom  225    96.470207

### 3) Which recipe, on average, is the best?
#### answer: recipe b

In [178]:
# Get the average of taste score grouped by recipe
df_cake_melt.groupby('recipe').taste_score.mean()

recipe
a    63.922201
b    76.736074
c    75.874748
d    62.864844
Name: taste_score, dtype: float64

In [ ]:
#### Recipe B has the best average taste score

### 4) Which oven temperature, on average, produces the best results?
#### answer: 275

In [118]:
# get the aveage of taste scores grouped by temperatures
df_cake_melt.groupby('temp').taste_score.mean()

temp
225    71.306022
250    66.577437
275    74.886754
300    66.627655
Name: taste_score, dtype: float64

#### Temperature 275 is the best 

### 5) Which combination of recipe, rack position, and temperature gives the best result?
#### answer: recipe b, bottom rack, 300 degrees

In [139]:
df_cake_combo = df_cake_melt.groupby(['recipe','rack_position','temp']).taste_score.mean()
df_cake_combo

recipe  rack_position  temp
a       bottom         225     61.738655
                       250     53.912627
                       275     74.414730
                       300     98.786784
        top            225     51.709751
                       250     52.009735
                       275     68.576858
                       300     50.228470
b       bottom         225     57.095320
                       250     61.904369
                       275     61.196980
                       300     99.248541
        top            225     82.455004
                       250     95.224151
                       275     98.594881
                       300     58.169349
c       bottom         225     96.470207
                       250     52.001358
                       275     92.893227
                       300     65.473084
        top            225     71.306308
                       250     82.795477
                       275     92.098049
                       300   

In [140]:
best_combo = df_cake_melt.groupby(['recipe','rack_position','temp']).taste_score.mean().max()
best_combo

99.2485405378462

In [143]:
df_cake_melt[df_cake_melt.taste_score == best_combo] 

recipe rack_position temp  taste_score
26      b        bottom  300    99.248541

#### # Best Combo is Recipe B, cooked in the bottom rack at 300 degrees